src/sqlast/visit.rs

```diff
@@ -0,0 +1,1284 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! SQL AST traversal.
+
+// Visiting an enum struct variant with many arguments forces our hand here.
+// If, over time, we convert these large variants to dedicated structs, we
+// can remove this escape hatch.
+#![allow(clippy::too_many_arguments)]
+
+// Disable lints that want us to rewrite `&SQLIdent` as `&str`, as `&str` is not
+// as self-documenting as &SQLIdent.
+#![allow(clippy::ptr_arg)]
+
+use super::*;
+
+/// A trait that represents a visitor that walks through a SQL AST.
+///
+/// Each function corresponds to a node in the SQL AST, and has a default
+/// implementation that visits all of its child nodes. Implementors of this
+/// trait can override functions as desired to hook into AST traversal without
+/// writing code to traverse the entire AST.
+pub trait Visit<'ast> {
+    fn visit_statement(&mut self, statement: &'ast SQLStatement) {
+        visit_statement(self, statement)
```

Comment:
I'm confused by the default implementation ... it looks like it recurses straight back into itself. I'm sure I'm missing something obvious here but could you explain this?

Reply:
Sure thing! I stole this pattern from the syn crate's visitor. The goal is to minimize the amount of code that the consumer needs to write.

Basically, there's a trait method with a default implementation that calls a free function (with the same name, hence the confusion), and that free function that does the actual walking of the AST. Here's the first few lines of `visit_statement`, for example:

```rust
pub fn visit_statement<'ast, V: Visit<'ast> + ?Sized>(
     visitor: &mut V,
     statement: &'ast SQLStatement,
 ) {
     match statement {
         SQLStatement::SQLQuery(query) => visitor.visit_query(query),
         SQLStatement::SQLInsert {
             table_name,
             columns,
             source,
         } => visitor.visit_insert(table_name, columns, source),
         // ...
     }
 }
```

Here's a more concrete example of a visitor that makes use of the free function avoid reimplementing the AST-walking code:

```rust
struct MyVisitor {
    // arbitrary state
}

impl<'ast> Visit<'ast> for MyVisitor<'ast> {
    fn visit_function(&mut self, func: &'ast SQLFunction) {
        // pre-visit
        sqlparser::visit::visit_function(func);
        // post-visit
    }
}
```

Basically, overriding trait methods is really convenient, because you only have to override the methods whose behavior you want to change. But sometimes you don't want to override the method entirely; you just want to tack on some behavior to the beginning or end. Having a public free function makes this quite ergonomic. You override the trait method, do whatever you need to, and then call the code to keep recursion when you're ready—or perhaps not at all, if you've already found what you need.

Reply:
Got it. Thanks for explaining that to me.